src/HealthChecks/HealthChecks/src/DefaultHealthCheckService.cs

```diff
@@ -18,36 +19,83 @@ namespace Microsoft.Extensions.Diagnostics.HealthChecks;
+    private readonly IOptions<HealthCheckPublisherOptions> _healthCheckPublisherOptions;
     private readonly ILogger<DefaultHealthCheckService> _logger;
+    private readonly ConcurrentQueue<KeyValuePair<string, HealthReportEntry>> _healthReportEntriesQueue;
+    private readonly IReadOnlyDictionary<TimeSpan, List<HealthCheckRegistration>> _periodHealthChecksMap;
+    private readonly ICollection<Timer> _healthCheckRegistrationTimers;
 
     public DefaultHealthCheckService(
         IServiceScopeFactory scopeFactory,
-        IOptions<HealthCheckServiceOptions> options,
+        IOptions<HealthCheckServiceOptions> healthCheckServiceOptions,
+        IOptions<HealthCheckPublisherOptions> healthCheckPublisherOptions,
         ILogger<DefaultHealthCheckService> logger)
     {
         _scopeFactory = scopeFactory ?? throw new ArgumentNullException(nameof(scopeFactory));
-        _options = options ?? throw new ArgumentNullException(nameof(options));
+        _healthCheckServiceOptions = healthCheckServiceOptions ?? throw new ArgumentNullException(nameof(healthCheckServiceOptions));
+        _healthCheckPublisherOptions = healthCheckPublisherOptions ?? throw new ArgumentNullException(nameof(healthCheckPublisherOptions));
         _logger = logger ?? throw new ArgumentNullException(nameof(logger));
 
         // We're specifically going out of our way to do this at startup time. We want to make sure you
         // get any kind of health-check related error as early as possible. Waiting until someone
         // actually tries to **run** health checks would be real baaaaad.
-        ValidateRegistrations(_options.Value.Registrations);
+        ValidateRegistrations(_healthCheckServiceOptions.Value.Registrations);
+
+        _healthReportEntriesQueue = new ConcurrentQueue<KeyValuePair<string, HealthReportEntry>>();
+
+        _periodHealthChecksMap = (from r in _healthCheckServiceOptions.Value.Registrations
+                                  where ByPredicate(r, _healthCheckPublisherOptions.Value.Predicate) // Apply publisher predicate
+                                  where r.Period != Timeout.InfiniteTimeSpan // Skip HCs with no individual period
+                                  group r by r.Period).ToDictionary<IGrouping<TimeSpan, HealthCheckRegistration>, TimeSpan, List<HealthCheckRegistration>>(g => g.Key, g => g.ToList());
+
+        _healthCheckRegistrationTimers = _periodHealthChecksMap.AsParallel().Select(t =>
+            NonCapturingTimer.Create(
+                async (state) =>
+                {
+                    var entries = await RunChecksAsync(t.Value).ConfigureAwait(false);
+                    foreach (var entry in entries)
+                    {
+                        _healthReportEntriesQueue.Enqueue(entry);
+                    }
+                },
+                null,
+                dueTime: healthCheckPublisherOptions.Value.Delay,
```

Comment:
Shouldn't they be using their own period rather than the publisher's?

Reply:
This is already the case: `period` param is passed the `t.Key` of the Dictionary, i.e. the TimeSpan period of the individual HC. `dueTime` param is passed the delay of the Publisher instead – unless we want to introduce a delay at the `HealthCheckRegistration` level as well?